### PR TITLE
Added burning trees to the Tiberian Dawn mod

### PR DESCRIPTION
**Pull request by Mailaender**
:evergreen_tree: can catch :fire: but only with flame inducing weaponry. They leave a scorched charcoaled tree, but can't be chopped down completely. Artwork for destructible fauna is available, but I fear the balancing shit storm as maps have been designed with trees as indestructible barriers in mind. Closes https://github.com/OpenRA/OpenRA/issues/6403. I can set up RA :deciduous_tree: as well if this gets positive feedback.

PS: Added an additional lint test to catch my own copy & paste mistakes which will crash later on: https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Buildings/FootprintUtils.cs#L51

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits.Render
+{
+	[Desc("Renders an overlay when the actor is taking heavy damage.")]
+	public class WithDamageOverlayInfo : ITraitInfo, Requires<RenderSpritesInfo>
+	{
+		public readonly string Image = "smoke_m";
+
+		[SequenceReference("Image")] public readonly string IdleSequence = "idle";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: Shouldn't this be called `StartSequence`?

File: mods/cnc/rules/defaults.yaml [hunk 2]
```diff
@@ -626,10 +626,40 @@
...
 
+^TreeHusk:
+	Inherits@1: ^SpriteActor
+	AppearsOnRadar:
+	Building:
+		Footprint: __ x_
+		Dimensions: 2,2
+	WithSpriteBody:
+	Tooltip:
+		Name: Tree (Destroyed)
```
Comment (pchote) [inline mods/cnc/rules/defaults.yaml]: IMO polish-wise it would be better to not include the "(Destroyed)" in the tooltip.

**Comment by pchote** [comment]
I haven't checked/tested all the details yet, but at a quick view this looks fine.  Great job!

We'll eventually (not here) want to replace the fire overlay with a proper implementation of #9409.

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: Technically it is Start + Loop. Somehow inherited from the existing implementation.

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+			if (isSmoking) return;
+			if (e.Damage < 0) return;	/* getting healed */
+			if (e.DamageState < info.MinimumDamageState) return;
+			if (e.DamageState > info.MaximumDamageState) return;
+
+			isSmoking = true;
+			anim.PlayThen(info.IdleSequence,
+				() => anim.PlayThen(info.LoopSequence,
+					() => anim.PlayBackwardsThen(info.EndSequence,
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: Why backwards?

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: It's carried over from `WithSmoke` though so I guess nonrelevant.

File: mods/cnc/rules/defaults.yaml (hunk 2 above)
Comment (GraionDilach) [inline mods/cnc/rules/defaults.yaml]: (Burnt) or (Dead) might work though.

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: That explains the glitches I was observing. Fixed it as playing this sequence backwards isn't obvious to modders.

**Comment by RoosterDragon** [comment]
Needs a rebase.

**Comment by Mailaender** [comment]
Rebased.

**Comment by GraionDilach** [comment]
>Checking for code style violations in OpenRA.Mods.Common...
>OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:L780: [SA1507] The code must not contain multiple blank lines in a row.

**Comment by Mailaender** [comment]
Hate this rebase conflicts. Should be fixed.

**Comment by pchote** [comment]
Looks fine to me now.  We can polish this further in followup PRs. :+1:

**Comment by GraionDilach** [comment]
>We'll eventually (not here) want to replace the fire overlay with a proper implementation of #9409.

Enable a SelfHealing upgrade (which heals negatively) based on that damage level trait @reaperrr sent in the other day, with this trait for visuals, done. :P

**Comment by pchote** [comment]
Not quite.  The fire position also needs to be randomized, and fires should spread to adjacent trees under certain circumstances.

**Comment by obrakmann** [comment]
:+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/e20e321bd74a767c2bd40cb49c002bdec1b3cb1c)

**Comment by Mailaender** [comment]
Expanded [changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e20e321bd74a767c2bd40cb49c002bdec1b3cb1c...91877549f9069f2c9e72de7155aeda6890c2bf75).

**Pull request merged**